### PR TITLE
[CELEBORN-860][DOC] Document on ShuffleClient

### PR DESCRIPTION
**Pull request by waitinfuture**
### What changes were proposed in this pull request?
As title.

### Why are the changes needed?
As title.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manual test.

### CODE REVIEW

**Comment by waitinfuture** [comment]
cc @pan3793 @cfmcgrady @AngersZhuuuu

File: docs/developers/shuffleclient.md [hunk 1]
```diff
@@ -111,4 +115,62 @@ The process of `SOFT_SPLIT` is as follows:
...
+- `endMapIndex` is the index of end map index of interested map range, set to a very large value if you want
+  to read all partition data
+
+The returned input stream is guaranteed to be `Exactly Once`, meaning no data lost and no duplicated reading, or else
+an exception will be thrown, see [Here](../../developers/faulttolerant#exactly-once).
+
+### Get PartitionLocations
+To read data from a partition id, `ShuffleClient` first checks whether the mapping from partition id to all
+`PartitionLocation`s are locally cached, if not, `ShuffleClient` sends GetReducerFileGroup to `LifecycleManager`
+for the mapping, see [Here](../../developers/lifecyclemanager#).
```
Comment (pan3793) [inline docs/developers/shuffleclient.md:147]: empty anchor?

**Review by pan3793 — APPROVED** [review]

File: docs/developers/shuffleclient.md [hunk 2]
```diff
@@ -111,4 +115,62 @@ The process of `SOFT_SPLIT` is as follows:
...
-to guarantee no data is lost.
\ No newline at end of file
+to guarantee no data is lost.
+
+## Read Data
+### API specification
+`ShuffleClient` provides an API that creates an InputStream to read data from a partition id. Users can also set
+`startMapIndex` and `endMapIndex` to read data within the map range.
+```java
+  public abstract CelebornInputStream readPartition(
```
Comment (cfmcgrady) [inline docs/developers/shuffleclient.md:125]: let's update the function comment as well.

File: docs/developers/shuffleclient.md (hunk 1 above)
Comment (waitinfuture) [reply docs/developers/shuffleclient.md]: fixed

File: docs/developers/shuffleclient.md (hunk 2 above)
Comment (waitinfuture) [reply docs/developers/shuffleclient.md:125]: done

Comment (cfmcgrady) [reply docs/developers/shuffleclient.md:125]: `pushData`?

Comment (waitinfuture) [reply docs/developers/shuffleclient.md:125]: done

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java [hunk 3]
```diff
@@ -100,6 +100,21 @@ public static FileSystem getHdfsFs(CelebornConf conf) {
   public abstract void setupLifecycleManagerRef(RpcEndpointRef endpointRef);
 
   // Write data to a specific reduce partition
+
+  /**
+   * @param shuffleId the unique shuffle id of the application
```
Comment (cfmcgrady) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:105]:
```suggestion
  /**
   * Write data to a specific reduce partition
   *
   * @param shuffleId the unique shuffle id of the application
```

**Review by cfmcgrady — APPROVED** [review]

**Comment by waitinfuture** [comment]
Thanks, merging to main.

**Pull request closed without merging**